### PR TITLE
Python bindings for EclIO classes

### PR DESCRIPTION
**Pull request by tskille**
This PR include python bindings for Eclipse IO classes EclFile, ERst, EGrid, ERft, ESmry and EclOutput

### CODE REVIEW

File: python/cxx/eclipse_io.cpp [hunk 1]
```diff
@@ -13,26 +13,75 @@ namespace py = pybind11;
...
+size_t get_array_index(const std::vector<EclEntry>& array_list, const std::string& array_name, size_t occurence)
+{
+    std::vector<std::string> arrayNameList;
+
+    arrayNameList.reserve(array_list.size());
+
+    for (auto element : array_list)
+        arrayNameList.push_back(std::get<0>(element));
+
+    std::vector<std::string>::iterator it = arrayNameList.begin();
```
Comment (akva2) [inline python/cxx/eclipse_io.cpp]: let's avoid all the copies.

```
    size_t cidx = 0;
    auto it = std::find_if(array_list.begin(), array_list.end(),
                           [&cidx, &array_name, occurence](const EclEntry& entry)
                           {
                              if (std::get<0>(entry) == array_name)
                                  ++cidx;

                              return cidx == occurence + 1;
                           });

    return std::distance(array_list.begin(), it);
```

File: opm/io/eclipse/ERft.hpp [hunk 2]
```diff
@@ -63,18 +67,24 @@ class ERft : public EclFile
...
+    std::map<int, std::tuple<int,int>> arrIndexRange;
     int numReports;
     std::vector<float> timeList;
 
     std::set<std::string> wellList;
     std::set<RftDate> dateList;
     RftReportList rftReportList;
 
-    std::map<std::pair<std::string,RftDate>,int> reportIndex;  //  mapping report index to wellName and date (tupe)
+    std::map<std::tuple<std::string,RftDate>,int> reportIndex;  //  mapping report index to wellName and date (tupe)
```
Comment (akva2) [inline opm/io/eclipse/ERft.hpp]: this member shadows with every single method taking reportIndex as a parameter. rename it, e.g. 'reportIndices', or change the parameter names in the methods.

File: opm/io/eclipse/ERst.hpp [hunk 3]
```diff
@@ -51,6 +51,7 @@ class ERst : public EclFile
     }
 
     int count(const std::string& name, int reportStepNumber) const; 
+    int numberOfReportSteps() const { return seqnum.size(); };
```
Comment (akva2) [inline opm/io/eclipse/ERst.hpp]: seqnum is  a vector, it's size is size_t, not int. i do not see see why we would want make it signed and truncated.

File: opm/io/eclipse/ESmry.hpp [hunk 4]
```diff
@@ -41,10 +41,14 @@ class ESmry
 
     std::vector<float> get_at_rstep(const std::string& name) const;
 
+    const std::vector<int>& get_startdat() const { return startdat; }
+
     const std::vector<std::string>& keywordList() const { return keyword; }
 
     int timestepIdxAtReportstepStart(const int reportStep) const;
 
+    int numberOfTimeSteps() const { return param[0].size(); }
```
Comment (akva2) [inline opm/io/eclipse/ESmry.hpp]: why int?

File: python/cxx/eclipse_io.cpp [hunk 5]
```diff
@@ -13,29 +20,268 @@ namespace py = pybind11;
 
 namespace {
 
-py::array get_vector(Opm::EclIO::EclFile * file_ptr, std::size_t array_index) {
+using npArray = std::tuple<py::array, Opm::EclIO::eclArrType>;
+using EclEntry = std::tuple<std::string, Opm::EclIO::eclArrType, long int>;
```
Comment (akva2) [inline python/cxx/eclipse_io.cpp]: any reason why this is a long int? it's icky for portability.

Comment (tskille) [reply python/cxx/eclipse_io.cpp]: As part of supporting huge arrays , PR #1587 "Adding support for larger EclFile arrays ", EclEntry was changed in EclFile.hpp from int to long int

```C++
using EclEntry = std::tuple<std::string, eclArrType, long int>;
```

Comment (akva2) [reply python/cxx/eclipse_io.cpp]: well as discussed in email windows has a surprise for you. it's better to use int64_t.

File: opm/io/eclipse/ERst.hpp (hunk 3 above)
Comment (bska) [reply opm/io/eclipse/ERst.hpp]: > it's size is size_t,

It actually isn't.  The return value from `vector<T>::size()` is really
```C++
std::vector<T>::size_type
```
which **happens to be** `std::size_t` in many implementations.

File: python/cxx/eclipse_io.cpp (hunk 5 above)
Comment (tskille) [reply python/cxx/eclipse_io.cpp]: I'm planning to go through the EclIO classes and replace long int with int64_t and from unsigned long int with uint64_t.  

I will preferre to do this in a separate PR. I will start with this when this is completed.

File: python/cxx/eclipse_io.cpp (hunk 1 above)
Comment (tskille) [reply python/cxx/eclipse_io.cpp]: done

File: opm/io/eclipse/ERft.hpp (hunk 2 above)
Comment (tskille) [reply opm/io/eclipse/ERft.hpp]: done

File: opm/io/eclipse/ERst.hpp (hunk 3 above)
Comment (tskille) [reply opm/io/eclipse/ERst.hpp]: done

File: opm/io/eclipse/ESmry.hpp (hunk 4 above)
Comment (tskille) [reply opm/io/eclipse/ESmry.hpp]: done

**Comment by tskille** [comment]
Thanks @akva2. I have now addressed all comments you had and hopefully this is now ready to be merge. When This is merged, I'll start working on replacing **long int** with **int64_t** and **unsigned long int** with **uint64_t**.

**Comment by akva2** [comment]
jenkins build this please

**Comment by akva2** [comment]
quell https://ci.opm-project.org/job/opm-common-PR-builder/1644/gcc/new/

**Comment by akva2** [comment]
jenkins build this please

**Comment by akva2** [comment]
https://ci.opm-project.org/job/opm-common-PR-builder/1669/gcc/new/source.0e9d384b-6277-4427-b662-7ad9d15ec884/#125

**Comment by akva2** [comment]
jenkins build this please

**Pull request merged**